### ISSUE TITLE
[Android] Random MediaPlayer IllegalStateException

### ISSUE THREAD

**Issue by FlammableFork** [adrianstevens/Xamarin-Plugins#9]
I am handling the PlaybackEnded event to know when my audio is finished playing.  Most of the time this works perfectly.  However, I have been experiencing an issue where randomly, and out-of-the-blue, the call to player.Prepare() at SimpleAudioPlayerImplementation.cs:232 is throwing an IllegalStateException.  Because this is in the SimpleAudioPlayerImplementation event handler, I cannot simply catch the exception, and it causes my app to terminate.  After investigating, it appears to be a timing issue where for some reason the player has already been torn down and rebuilt before the event handler is finished processing.

I have tried both reusing the same ISimpleAudioPlayer repeatedly as well as constructing a new instance each time I need it (I am playing several short tones, possibly in quick succession, but always waiting until after the previous tone is finished playing).  Both approaches result in the same outcome; at some point the IllegalStateException is thrown on line 232.  It may occur after playing the first tone, or it may occur after playing the 15th; it appears to be very random.

I have played with code in the repository, and I believe I have found a way around this problem by reusing the player instead of rebuilding it.  Would you mind taking a look at the diff below to see what I've come up with and see if you think this is a workable approach for you?  While I was testing with these changes, I was unable to cause the IllegalStateException any longer, no matter what I tried.  Thanks.

```diff
--- SimpleAudioPlayerImplementation.cs
+++ SimpleAudioPlayerImplementation.cs
@@ -79,9 +79,9 @@
+       /// <summary>
+       /// Instantiates a new SimpleAudioPlayer
+       /// </summary>
+       public SimpleAudioPlayerImplementation()
+       {
+            player = new Android.Media.MediaPlayer() { Looping = Loop };
+       }
+
        ///<Summary>
        /// Load wav or mp3 audio file as a stream
        ///</Summary>
        public bool Load(Stream audioStream)
        {
-           DeletePlayer();
+           player.Reset();

            //cache to the file system
            path = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.Personal), $"cache{index++}.wav");
            var fileStream = File.Create(path);
            audioStream.CopyTo(fileStream);
            fileStream.Close();

-           player = new Android.Media.MediaPlayer() { Looping = Loop };
-
            try
            {
                player.SetDataSource(path);
            }
            catch
            {
                try
                {
                    var context = Android.App.Application.Context;
                    player?.SetDataSource(context, Uri.Parse(Uri.Encode(path)));
                }
                catch
                {
                    return false;
                }
            }

            return PreparePlayer();
        }

        ///<Summary>
        /// Load wav or mp3 audio file from the iOS Resources folder
        ///</Summary>
        public bool Load(string fileName)
        {
-           DeletePlayer();
+           player.Reset();

            AssetFileDescriptor afd = Android.App.Application.Context.Assets.OpenFd(fileName);

-           player = new Android.Media.MediaPlayer() { Looping = Loop };
-
            player?.SetDataSource(afd.FileDescriptor, afd.StartOffset, afd.Length);

            return PreparePlayer();
        }
```

**Comment by ihassantariq** [comment]
Hey, I am having same issue. Just let me know how can I fix this in only android. I have checked UWP and IOS is working fine. There wasn't any issue with those platform. @FlammableFork  can you help me out with the implementation on Android platform.

**Comment by FlammableFork** [comment]
@ihassantariq I downloaded the source and built my own NuGet package in order to use it in my app.  I kept pretty much everything as is (including publishing information) except that I didn't/couldn't build the MacOS package, so I left it out.  I also incremented the version to 1.0.2-pre (to keep versioning close to the original).  Feel free to try my NuGet package and see if it works for you.  I installed this package in all of my projects just to keep the package version consistent.  Either way, things have been working flawlessly for me since.  Thanks.

[Xam.Plugin.SimpleAudioPlayer.1.0.2-pre7.zip](https://github.com/adrianstevens/Xamarin-Plugins/files/1583551/Xam.Plugin.SimpleAudioPlayer.1.0.2-pre7.zip)

If you aren't aware, or haven't done it before, you will need to add a new package source in order to install this package.  If using Visual Studio, go to 'Tools -> Options -> NuGet Package Manager -> Package Sources' and add a new source pointing to the location where you extracted the package.

Then, when you try to install the package, simply make sure that the package source is set to either 'All' or specifically the one you created/added.  Also make sure that 'Include prerelease' is SELECTED.

![image](https://user-images.githubusercontent.com/22598063/34313748-c2022f70-e72a-11e7-8b96-9683615279e4.png)

**Comment by ihassantariq** [comment]
I have tried your provided plugin implementation. In fact I have try to use the your plugin implementation. But problem I am having is that if I don't create the new instance then it is not working second time. If I try to use Plugin.SimpleAudioPlayer.CrossSimpleAudioPlayer.Current instance it works second time and never third time. Actually, I have to change source every time I play new tone or song whatever. I think it caches the path and on when Play() event is invoked internally it calls the Start() virtual function but I don't know why it does not work second time if I try to use old instance. Can you paste a little implementation on how you are using this! This would be very helpful. Or if you want to shed some light on my use case that would be okay too.

**Comment by FlammableFork** [comment]
@ihassantariq I too am using Plugin.SimpleAudioPlayer.CrossSimpleAudioPlayer.Current which always reuses the same instance every time.  In my app I am ensuring the previous audio is completely finished playing (I use the PlaybackEnded event) before allowing a new source to be used as I don't require pausing or stopping mid-playback.  Without knowing the specifics of your application, I can only think of two possibilities:

1. The player is still playing when you try to load a new source.
    * Try making certain the previous source is finished playing by calling CrossSimpleAudioPlayer.Current.Stop()
    * This shouldn't really be necessary as the Android media state diagram claims that reset() can be called in the 'Started' state.
2. The audio source you tried to load did not actually load successfully.
    * Are you using the return type from the CrossSimpleAudioPlayer.Current.Load() method to determine whether the source loaded properly?  The Load() method catches all exceptions while loading a source and simply returns false in the case of an error.  This won't help you determine the reason for the failure, but it would, at least, let you know that something failed.

Sorry I cannot be more help.  I am reusing the instance of the player as you are, and things are working well for me.

**Comment by FlammableFork** [comment]
@ihassantariq At the end of the day, I am doing nothing more than this:

```C#
private void PlayAudio(string filename) // Configure this however you need
{
	// Get Audio Player
	ISimpleAudioPlayer audioPlayer = CrossSimpleAudioPlayer.Current;

	// Stop Playback (I'm not doing this either, but it might make for a better transition, if necessary)
	audioPlayer.Stop();
	
	// Register Playback Complete Event Handler
	audioPlayer.PlaybackEnded += OnPlaybackComplete;
	
	// Load Audio
	if (audioPlayer.Load(filename))
	{
		// Play Audio
		audioPlayer.Play();
	}
	else
	{
		// Throw Exception (I'm not actually doing this, but it might help troubleshoot depending on your code)
		throw new Exception("Audio failed to load");
	}
}

private void OnPlaybackComplete(object sender, EventArgs e)
{
	// Remove Event Handler
	CrossSimpleAudioPlayer.Current.PlaybackEnded -= OnPlaybackComplete;

	// Do Other Processing, As Needed
}
```

**Comment by ihassantariq** [comment]
@FlammableFork Thank you very much. I have tried your approach. But As I am working in PCL without using  CrossSimpleAudioPlayer.CreateSimpleAudioPlayer() this function I am unable to work with Player second time. I don't know it does not play second time. I am thinking on moving to another AudioPlayer Implementation. 👍  But Thanks for helping me out here.

**Comment by FlammableFork** [comment]
@ihassantariq Sorry you are having trouble getting it to work.  For what it's worth, all of my code is in PCL (.NET Standard library to be honest), and I am using CrossSimpleAudio.Current as you are.  Have you made sure that the second source is actually loading?  The plugin does work.  You also don't have to use the code I provided inside of methods or even at all.  You can use it however you need to (not knowing what your code looks like).  Regardless of what you choose to do, I wish you luck..

**Comment by ihassantariq** [comment]
Okay. Just let me give this another try. Let me remove the old plugin and used yours and I will give you my feedback regarding working of second time. If thats not work I will paste some of code here. Maybe you can help me figure out why I am having this type of behavior. And yes I have checked and made sure that Source loaded correctly.

**Comment by ihassantariq** [comment]
Nope. It is not running playing second time using your implementation of plugin. 
Let me paste important methods here
// First time initialization 
void InitailizeAudioPlayer()
{
			var stream = GetStreamFromFile(songs[currentIndex].SongName);

			player = CrossSimpleAudioPlayer.Current;
			player.PlaybackEnded += PlayBackEnded;
			player.Load(stream);
			player.Play();
}
//getting file from PCL project 
Stream GetStreamFromFile(string filename)
{
			var assembly = typeof(App).GetTypeInfo().Assembly;

			var stream = assembly.GetManifestResourceStream("SimVoice.Audios." + filename);

			return stream;
}

//Play back implementation 
	void PlayBackEnded(object sender, EventArgs args)
		{
			player = CrossSimpleAudioPlayer.Current;
			player.PlaybackEnded += PlayBackEnded;

                       // this function handles playing second time 
                         IncrementAndPlay();
                        //ignore these functions
			RemovePlayIconFromGrid();
			
			MakeCurrentGridItemToChangePhoto();
		}
// Playing successively
void IncrementAndPlay()
		{
			//before incrimenting lets make is play currently played element to false
			songs[currentIndex].IsPlaying = false;
			++currentIndex;

			if (currentIndex == 16)
			{
				currentIndex = 0;
			}
			// lets make is play currently played element to true
			songs[currentIndex].IsPlaying = true;


			var stream = GetStreamFromFile(songs[currentIndex].SongName);
			player = CrossSimpleAudioPlayer.Current;
			player.PlaybackEnded += PlayBackEnded;
			bool isLoaded = player.Load(stream);

			if (isLoaded)
			{
				player.Play();
			}
		}

Just let me know what I am doing wrong.

**Comment by ihassantariq** [comment]
@FlammableFork Also I am using the Nexus 4 emulator.

**Comment by FlammableFork** [comment]
@ihassantariq I see a one particular issue which would affect the third play and beyond, but nothing stands out that should affect the second play.

'player' is an instance variable, so it technically would need to be assigned to only once (perhaps in the constructor). You are assigning the same object to it 3 times (not a real issue).  The bigger issue there is that you are also registering your PlayBackEnded handler three times before the second file plays with an additional two times after each playback ends.  Each time you do this: player.PlaybackEnded += PlayBackEnded you are registering the event handler again, so when the second playback ends, you will fire the handler three times, which will completely mess up your currentIndex as well as anything else that happens in the handler.

Have you set a breakpoint at the final player.Play() inside IncrementAndPlay? Just to make sure that the player is going to try and play the stream?  I'm afraid that for the second stream to not be playing, the only thing I can think of would be that the loading of the second stream is failing for some reason, which would then simply not play.  Sorry.

**Comment by ihassantariq** [comment]
@FlammableFork I have solved it in Android. The solution is way more simpler than we think. I just have to invoke the Play() event on Device.InvokeOnMainThread. For debugging purpose I have added the wait for 1 second and in IncrimentAndPlay() function and it works. Then I realize this could be thread issue. So I have used Device.InvookeOnMainThread. And it is working flawlessly. Thank you for you help 👍

**Comment by adrianstevens** [comment]
@FlammableFork I'll be testing your changes this week and if all goes well I'll use it in the next update.

**Comment by adrianstevens** [comment]
@ihassantariq thank you posting your solution! Yes you'll need to play audio on the main/UI thread but that's not always obvious. I'll add a note to the readme.

**Comment by adrianstevens** [comment]
The code has been merged and version 1.0.2 is now released. I'm closing the issue but please feel free to re-open if further discussion is needed.

**Comment by BioTurboNick** [comment]
I just received an IllegalStateException for MediaPlayer from my published Android app using 1.1.0. Should I open a new issue?

**Comment by BioTurboNick** [comment]
I believe this a related issue. It is occurring in a call to the Prepare method which is being called in the OnCompletion event. This occurred on an Android 5.1 device,
